Prevent blocking loops on `PlainActionFuture`

DaveCTurner opened the pull request:
Extending the work of #108934, this commit adds a check that there are
no loops of executors that might block on each other, causing a
harder-to-detect possible deadlock than one that only involves a single
executor.

prdoyle commented on server/src/main/java/org/elasticsearch/action/support/AllowedExecutors.java:
```diff
@@ -0,0 +1,178 @@
+            newEntry(
+                "file-watcher",
+                /* blocks on futures completed by: */
+                "masterService#updateTask"
```
Could this be a record with two fields called `blocksOn` and `completedBy` or similar?

DaveCTurner replied:
Yeah ultimately we need to write down some pairs of strings somewhere, but e.g. `searchable_snapshots_cache_fetch_async` is really `org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots#CACHE_FETCH_ASYNC_THREAD_POOL_NAME` which is in `:x-pack:plugin:searchable-snapshots` so we can't write it here.

The question is more how to get those pairs of strings into this currently-statically-defined map. Something something SPI something something perhaps?

henningandersen replied:
Could this not be part of the executor builders that we return? I.e., `allowComplete` and `allowCompleteBy`? That would allow either direction to work I think.

DaveCTurner replied:
Yeah defining the pairs of strings somewhere like that should mostly work, it's just about how to get them into this currently-statically-defined map.

henningandersen replied:
I see. I wonder if we can make the map non-static and pass it as a thread local instead (noting that none of that will be used in prod, only with assertions enabled).

henningandersen replied:
(in fact, then it could be just the set of allowed threads to complete I think).

DaveCTurner committed:
Prevent blocking loops on `PlainActionFuture`

Extending the work of #108934, this commit adds a check that there are
no loops of executors that might block on each other, causing a
harder-to-detect possible deadlock than one that only involves a single
executor.

DaveCTurner commented:
FYI @henningandersen

DaveCTurner commented:
On reflection, even this check is insufficient. If we use a particular executor `A` for blocking on a future `F`, and `A` dispatches some work to executor `B` which ultimately triggers the completion of `F` on executor `C`, then we must also prevent `B` from blocking on futures completed by `A`. The check in this PR wouldn't prevent that.

I'm closing this because I don't think we're going to proceed with it in the near future.
